### PR TITLE
[SPARK-36575][CORE] Should ignore task finished event if its task set is gone in TaskSchedulerImpl.handleSuccessfulTask

### PR DESCRIPTION
**Pull request by sleep1661**
### What changes were proposed in this pull request?

When a executor finished a task of some stage, the driver will receive a `StatusUpdate` event to handle it. At the same time the driver found the executor heartbeat timed out, so the dirver also need handle ExecutorLost event simultaneously. There was a race condition issues here, which will make `TaskSetManager.successful` and `TaskSetManager.tasksSuccessful` wrong result.

The problem is that `TaskResultGetter.enqueueSuccessfulTask` use asynchronous thread to handle successful task, that mean the synchronized lock of `TaskSchedulerImpl` was released prematurely during midway https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala#L61. So `TaskSchedulerImpl` may handle executorLost first, then the asynchronous thread will go on to handle successful task. It cause `TaskSetManager.successful` and `TaskSetManager.tasksSuccessful` wrong result.

### Why are the changes needed?

 It will cause `TaskSetManager.successful` and `TaskSetManager.tasksSuccessful` wrong result. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Add a new test.

### CODE REVIEW

**Comment by sarutak** [comment]
@sleep1661 Could you rebase your repository to `master`?
Also, could you enable GitHub Actions on your account?

cc: @mridulm @Ngone51

**Comment by sleep1661** [comment]
@sarutak I have corrected it. Thanks.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:877]: nit: s"Ignoring update with state finished for TID $tid because ...

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:876]: Make this a `logWarn` or `logInfo` even ? This is not something users can mitigate or cause issues once we have handled it this way.

**Review by mridulm — COMMENTED** [review]
The fix looks reasonable, but I want to make sure @Ngone51, @tgravescs or someone familiar with scheduler also takes a look.

Orthogonal to this bug, looks like we are not consistently locking on `TaskSchedulerImpl` when using `taskIdToTaskSetManager` - thoughts @Ngone51 ?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2051]: Using thread.sleep's here will result in flakey tests.
Instead, mock the relevant methods and use latches to control behavior.

**Review by Ngone51 — COMMENTED** [review]
I don't understand how the issue here leads to the stage hang. From what I see now, the task is considered a failed task first (due to executor loss) and reenqueue to the pending list, then it's considered a successful task later (due to statusupdate event). This does cause some inconsistent status but doesn't seem to cause the stage hang. IIRC, the stage can also complete even if the pending list isn't empty.

I probably missing something here. Could you elaborate more on how it causes stage hang?

**Comment by Ngone51** [comment]
> Orthogonal to this bug, looks like we are not consistently locking on TaskSchedulerImpl when using taskIdToTaskSetManager - thoughts @Ngone51 ?

Looks so. I found 2 places:

https://github.com/apache/spark/blob/9b96b705b2cce403ba5339cd481648c19bec5147/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L856

https://github.com/apache/spark/blob/9b96b705b2cce403ba5339cd481648c19bec5147/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L371-L375


but both looks fine since `taskIdToTaskSetManager` itself is also a `ConcurrentHashMap`.

**Comment by Ngone51** [comment]
@sleep1661 btw, the PR looks referred to a wrong JIRA number. Could you correct it?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (sleep1661) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Ok, I will adjust that.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (sleep1661) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: `logInfo` will be fine.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: I had optimized it, you can review it if you have time.

Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: I had optimized it, you can review it if you have time.

> I don't understand how the issue here leads to the stage hang. From what I see now, the task is considered a failed task first (due to executor loss) and reenqueue to the pending list, then it's considered a successful task later (due to statusupdate event). This does cause some inconsistent status but doesn't seem to cause the stage hang. IIRC, the stage can also complete even if the pending list isn't empty.
> 
> I probably missing something here. Could you elaborate more on how it causes stage hang?

The issue will  cause `TaskSetManager.successful` wrong result, mark the partition index true.  And `TaskSetManager.dequeueTaskFromList` only dequeue the unsuccessful index (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L313), so this partition index will be never scheduled again.
I also had update unit test., add assert code `assert(false == taskSetManager.successful(taskDescriptions(0).index))`.

**Comment by sleep1661** [comment]
> @sleep1661 btw, the PR looks referred to a wrong JIRA number. Could you correct it?

Done. Thanks.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: > The issue will cause TaskSetManager.successful wrong result, mark the partition index true. And TaskSetManager.dequeueTaskFromList only dequeue the unsuccessful index (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L313), so this partition index will be never scheduled again.


As I mentioned, the stage can also complete even if the pending list isn't empty. I think it can still complete as long as `tasksSuccessful == numTasks`. I still can't get why it can lead to stage hang.

**Comment by mridulm** [comment]
Ok to test

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2027]: Both tasks have the same partition index - can you fix them to be different, and then get the testcase to surface the issue ?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Based on your question, I realize that some details are indeed missing. So reorganize the whole process based on the log. Finally, it was found that `TaskSetManager.executorLost` was executed twice, and the second time resulted in `tasksSuccessful -= 1`, resulting in `tasksSuccessful` always less than `numTasks` .
The reason `TaskSetManager.executorLost` is executed twice is because `Heartbeat.expireDeadHosts` executes `TaskSchedulerImpl.executorLost` for the first time, and then executes `sc.killAndReplaceExecutor`. This issue occurred in version 2.3.3. `Heartbeat.expireDeadHosts` had adjusted the logic of the master branch, so I am not sure whether the stage hang issue was existed in new version. BTW,  I updated the JIRA description, added more detail log.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: done

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Ok, I checked the logs in JIRA with the 2.3 code. Now I can guess how `TaskSetManager.executorLost` can be called twice. The first time is called by `TaskSchedulerImpl.executorLost` in `HeartbeatReceiver`, and  the second time could be triggered by `onDisconnect -> removeExecutor -> executorLost` due to `sc.killAndReplaceExecutor`. And the second time can only take effect when a `resourceOffer` request happens before it so that the `executorId` can be added back to `executorIdToHost`.  But, as you can see,

https://github.com/apache/spark/blob/75cc3b2da9ee0b51ecf0f13169f2b634e36a60c4/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L936-L949

`tasksSuccessful -= 1` only happens when `successful(index)=true`. And after first `executorLost`, `successful(index)` is already `false`. So, I think `tasksSuccessful -= 1` wouldn't happen the second time. 

I hope you can provide the complete or more logs if possible so that we can figure out the what's the real issue here. For now, I still can't see how the stage gets hang.

Beside, in Master, `TaskSchedulerImpl.executorLost` is no longer directly called in `HeartbeatReceiver`. Instead, it's delegated to the `CoarseGrainedSchedulerBackend`. And since `sc.killAndReplaceExecutor` would also be delegated to the `CoarseGrainedSchedulerBackend` at the end, thus, executor data would be synchronized by `CoarseGrainedSchedulerBackend`. So, I think call `executorLost` twice won't happen in Master.

Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: > `tasksSuccessful -= 1` only happens when `successful(index)=true`. And after first `executorLost`, `successful(index)` is already `false`. So, I think `tasksSuccessful -= 1` wouldn't happen the second time.
> 
There is a finished event of that task between the first `executorLost` and the second `executorLost`, so `successful(index)` was `true` when second `executorLost` was called.

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: > Finally, it was found that TaskSetManager.executorLost was executed twice, and the second time resulted in tasksSuccessful -= 1, resulting in tasksSuccessful always less than numTask.

Then, how could it result in "tasksSuccessful always less than numTask" if there's a task finish event? The event, IIUC, is a successful event, which will do `tasksSuccessful +=1` first, right?

Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: > 
>  The event, IIUC, is a successful event, which will do `tasksSuccessful +=1` first, right?

Yes. But the second `executorLost` made `tasksSuccessful -= 1`. With `successful(index)` was true now,  the task would never been scheduled again,  which resulted in "tasksSuccessful always less than numTask"

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Wouldn't the second `executorLost` reset `successful(index)` to false?

Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: You are right, `successful(index)` is indeed no problem, it is my mistake. But I also saw `copiesRunning(index) -= 1` were processed in `executorLost` , twice `executorLost`  made `copiesRunning(index)` to `-1`.  So this may be the real problem, am i right ?

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Ok, that's true. I think the root cause is due to `copiesRunning(index) = -1`. But the problem exists in 2.3. In Master, as I mentioned above, twice `executorLost` can no longer happen.

**Comment by mridulm** [comment]
@sleep1661 The test is not surfacing an issue in master - can you reformulate it such that the problem is surfaced in master without the patch and gets fixed after the patch ?
(Currently it is failing due to `tasksSuccessful` > 0 - which is not an error in itself).


@Ngone51 I was initially looking at the patch from point of view of inconsistent events being fired - (two events for the same task - first failed and then success) : and from that point of view, it looked like a right fix to have.
It is a good idea to investigate if the root cause reported is actually an issue - namely if we hang.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (sleep1661) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Yes. But it cause` TaskSetManager.successful` and `TaskSetManager.tasksSuccessful` wrong result in Matser, I think it's better to fix it.

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Yea, but could you rephrase the PR description first?

**Comment by Ngone51** [comment]
@mridulm As we discussed at https://github.com/apache/spark/pull/33872#discussion_r703279670, I think me and @sleep1661 have reached the agreement that the hang issue exists in 2.3 but not master.  So, now, I think our target is to fix the inconsistent status of the task(I'm actually not sure if the issue really exists because I didn't pay much attention on this before. I asked @sleep1661 to update the PR description first. So let's see.).  To clarify, I think Spark actually can heal the inconsistent issue by rerun the stage later. But if we could fix it early in TaskSetManager, then, we can fix the issue eariler with less cost.

**Comment by mridulm** [comment]
@Ngone51 Before this fix, we have the following three cases primarily:

1. Final ResultStage
    - These wont have a task re-execution as the result has already been fetched (collect/save/etc).
2. ShuffleMapStage with ESS enabled.
    - If the node has not gone down as well, but only executor has failed, then we wont have re-execution.
3. ShuffleMapStage without ESS enabled.
    - This will have re-execution as the shuffle blocks are lost.

In addition to above, if the stage was persisting data - that would also trigger re-execution when there is a failure to fetch the RDD blocks (assuming no replication, etc). Given this, we wont necessarily have stage re-execution in all cases in existing master - but only when there is need for data which is actually lost.
On the other hand, the task events would be inconsistent whenever the problem is hit.

After the change, we will re-execute the task for all of the cases above if this issue is hit (executor failure racing against successful result fetch).
I am actually fine with this change in behavior given the consistent way we handle events.

**Comment by Ngone51** [comment]
Sorry, I missed the conversation here.

@mridulm Yes, I also agree with the change.

@sleep1661 Again, could you update the PR description?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:38]: could you follow the imports style as others?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2043]: Move to the imports group

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2080]:
```suggestion
    assert(!taskSetManager.successful(taskDescriptions(0).index))
```

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:40]:
```suggestion
import org.apache.spark.util.{Clock, ManualClock, ThreadUtils}
```

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:876]: `....for TID $tid...` -> `for task (TID $tid)...`

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:877]: I wonder is it possible to receive "duplicate task finished status updates"?

Comment (sleep1661) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:878]: I'm not sure. I just infer based on this log (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L825)

**Comment by sleep1661** [comment]
Ping @Ngone51

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:878]: sounds good.

**Comment by Ngone51** [comment]
Regarding "There was a race condition issues here, which will make TaskSetManager.successful and TaskSetManager.tasksSuccessful inconsistent." in the PR description, could you explain more?  It's been a while since the last time I looked into this issue, and I currently can't see how `successful` and `tasksSuccessful` become inconsistent.

**Comment by sleep1661** [comment]
> Regarding "There was a race condition issues here, which will make TaskSetManager.successful and TaskSetManager.tasksSuccessful inconsistent." in the PR description, could you explain more? It's been a while since the last time I looked into this issue, and I currently can't see how `successful` and `tasksSuccessful` become inconsistent.

You can take a look at my unit test. Without my PR, the `tasksSuccessful` will become  `1` and `successful`  will be `true`

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (sleep1661) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: done.

**Review by Ngone51 — APPROVED** [review]

**Comment by Ngone51** [comment]
@mridulm do you have more concerns? I'll merge this if you're also ok with it.

**Review by mridulm — APPROVED** [review]

**Comment by Ngone51** [comment]
Thanks, merged to Master!

**Comment by Ngone51** [comment]
@sleep1661 Is ` hujiahua` your jira account id? I can't find you to assign you the ticket there.

**Comment by sleep1661** [comment]
> @sleep1661 Is ` hujiahua` your jira account id? I can't find you to assign you the ticket there.

No, `sleep1661` is my jira account id

**Comment by sarutak** [comment]
@Ngone51 @sleep1661 sleep1661 is the `Username`. your account id is `hujiahua`.
https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sleep1661

**Comment by Ngone51** [comment]
Thanks @sarutak . I tried `hujiahua` but jira told me there's no matches. It's weird.

**Comment by sarutak** [comment]
@Ngone51 
Ah, you might need to add him as a contributor.
https://issues.apache.org/jira/plugins/servlet/project-config/SPARK/roles

**Comment by Ngone51** [comment]
Ah, it works now! Thanks for the help @sarutak !!

**Comment by ulysses-you** [comment]
sorry, but it seems something wrong after merged this PR. Every test is running without finish. And alos I cannot find a success pass with jenkins in this PR. Do you have time to take a look ? @Ngone51

**Comment by Ngone51** [comment]
oh..thanks for the report..let's me take a quick look.

**Comment by Ngone51** [comment]
I've reverted the commit. @sleep1661 Could take a look at the test issue?

**Comment by sleep1661** [comment]
It’s a bit weird, my changes are quite few and I really didn’t see where the problem was. Any idea ? @Ngone51

**Comment by Ngone51** [comment]
It turns out that we always cleanup the `taskIdToTaskSetManager` for a finished task via `cleanupTaskState(tid)`:

https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L811-L815

Therefore, `taskIdToTaskSetManager.contains(tid)` would always return `false` in `handleSuccessfulTask`, which leads to no task can update its finished state normally.

**Comment by Ngone51** [comment]
Actually, `taskIdToTaskSetManager` is still under the protection of `TaskSchedulerImpl`'s lock. `taskIdToTaskSetManager` has no problem here. The real problem is the thread-safety of `TaskSetManager.taskInfos`. Ideally, as for this issue, `handleSuccessfulTask` should return immediately when `taskInfo.finished=true` (caused by executorLost). That said, we should be careful to address thread-safety issue of `taskInfo` since it's referenced at the hot code path. I'd like to investigate more to see if we have other better solutions.

**Comment by sleep1661** [comment]
You are right.  We need think other better solutions.

**Comment by Ngone51** [comment]
https://github.com/apache/spark/blob/9b96b705b2cce403ba5339cd481648c19bec5147/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L47-L48

Actually, all the states in `TaskSetManager` are already protected by `TaskSchudlerImpl`. So, `TaskSetManager.taskInfos` is actually thread-safe. I also confirmed it by checking the call chain around `handleSuccessfulTask` and `executorLost`. Therefore, we don't need extra protection on the `TaskSetManager.taskInfos`.

@sleep1661 Would you like to continue working on this? Otherwise, I could take it over.
BTW, I think we need a separate JIRA ticket for it since the real issue isn't the hanging issue and fix in a separate PR as well.

**Comment by sleep1661** [comment]
@Ngone51 Sure,  I will go on working on this. I will create a new JIRA ticket and new PR this week,  and thank you for your code investigation and analysis.

**Comment by mridulm** [comment]
Looks like the jenkins tests had failed with timeout - I thought GA tests had passed, but I could be misremembering.
This is my bad, should have checked better.

**Comment by Ngone51** [comment]
My bad actually. Sorry about that.

**Comment by sleep1661** [comment]
@Ngone51 @mridulm  I  had created new PR (https://github.com/apache/spark/pull/34578) for this.  BTW sorry about the unit test problems caused by this PR.

**Pull request closed without merging**
